File: python3/69.sqrt-x.py
#
# @lc app=leetcode id=69 lang=python3
#
# [69] Sqrt(x)
#
# https://leetcode.com/problems/sqrtx/description/
#
# algorithms
# Easy (36.33%)
# Likes:    3476
# Dislikes: 3063
# Total Accepted:    979.4K
# Total Submissions: 2.7M
# Testcase Example:  '4'
#
# Given a non-negative integer x, compute and return the square root of x.
# 
# Since the return type is an integer, the decimal digits are truncated, and
# only the integer part of the result is returned.
# 
# Note: You are not allowed to use any built-in exponent function or operator,
# such as pow(x, 0.5) or x ** 0.5.
# 
# 
# Example 1:
# 
# 
# Input: x = 4
# Output: 2
# 
# 
# Example 2:
# 
# 
# Input: x = 8
# Output: 2
# Explanation: The square root of 8 is 2.82842..., and since the decimal part
# is truncated, 2 is returned.
# 
# 
# Constraints:
# 
# 
# 0 <= x <= 2^31 - 1
# 
# 
#

# @lc code=start
class Solution:
    def mySqrt(self, x: int) -> int:
        lo, hi = 1, x
        while lo < hi:
            mid = (lo + hi) // 2
            if mid*mid < x: lo = mid+1
            else: hi = mid
        return lo if lo*lo == x else lo-1
        
# @lc code=end

File: java/69.sqrt-x.java
/*
 * @lc app=leetcode id=69 lang=java
 *
 * [69] Sqrt(x)
 */

// @lc code=start
class Solution {
    public int mySqrt(int x) {
        int l = 0, r = x;
        while (l < r) {
            int mid = l + (r - l) / 2;
            long square = (long) mid * mid;
            if (square <= x) l = mid + 1;
            else r = mid;
        }

        return (l == 0 || l == 1) ? l : l-1;
    }
}
// @lc code=end

